### PR TITLE
ISPN-11575 EncryptedPropertiesSecurityRealm

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-11575

### CODE REVIEW

File: server/runtime/src/main/java/org/infinispan/server/logging/Log.java [hunk 1]
```diff
@@ -155,4 +157,13 @@
 
    @Message(value = "Cannot find InitialContextFactory '%s'", id = 80039)
    NoInitialContextException noInitialContextFactory(String className);
+
+   @Message(value = "Invalid Unicode sequence '%s'", id = 80040)
+   IOException invalidUnicodeSequence(String sequence, @Cause NoSuchElementException e);
+
+   @Message(value = "Decoding hashed password from users property file failed - should not be set as plain-text property file?", id = 80041)
+   RealmUnavailableException decodingPasswordFromPropertiesRealmFailed(@Cause Exception e);
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/logging/Log.java:165]: Unused Exception.

File: server/runtime/src/main/java/org/infinispan/server/security/UserTool.java [hunk 2]
```diff
@@ -28,22 +41,47 @@
...
+
    protected void handleArgumentCommand(String command, String parameter, Iterator<String> args) {
       switch (command) {
+         case "-a":
+            parameter = args.next();
+         case "--algorithms":
+            algorithms = new ArrayList<>();
+            for (String algorithm : parameter.split(",")) {
+               algorithms.add(algorithm);
+            }
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/security/UserTool.java:83]: `algorithms = Arrays.stream(parameter.split(",")).collect(Collectors.toList());`

File: server/runtime/src/main/java/org/infinispan/server/security/UserTool.java [hunk 3]
```diff
@@ -166,29 +209,55 @@ protected void runInternal() {
...
+                  break;
+               case DigestPassword.ALGORITHM_DIGEST_MD5:
+               case DigestPassword.ALGORITHM_DIGEST_SHA:
+               case DigestPassword.ALGORITHM_DIGEST_SHA_256:
+               case DigestPassword.ALGORITHM_DIGEST_SHA_384:
+               case DigestPassword.ALGORITHM_DIGEST_SHA_512:
+                  spec = new DigestPasswordAlgorithmSpec(username, realm);
+                  break;
+               default:
+                  throw new IllegalArgumentException("Unknown algorithm " + algorithm);
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/security/UserTool.java:235]: Should we create this String using `Messages.java`?

File: server/runtime/src/main/java/org/infinispan/server/security/UserTool.java [hunk 4]
```diff
@@ -151,11 +193,12 @@ protected void runInternal() {
             exit(1);
          }
       }
-      users.put(username, plainText ? password : hashPassword(username, password, realm));
+      users.put(username, plainText ? password : encryptPassword(username, realm, password));
       groups.put(username, addGroups.stream().collect(Collectors.joining(",")));
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/security/UserTool.java:197]: `groups.put(username, String.join(",", addGroups));`

File: server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+   private static final String COMMENT_SUFFIX = "$";
+   private static final String ALGORITHM_COMMENT_PREFIX = "$ALGORITHM=";
+
+   private final Supplier<Provider[]> providers;
+   private final String defaultRealm;
+   private final boolean plainText;
+
+   private final String groupsAttribute;
+
+   private final AtomicReference<LoadedState> loadedState = new AtomicReference<>();
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java:68]: Unnecessary whitespace between vars.

File: server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+                  return credentialType.cast(credential.clone());
+               }
+            }
+            return null;
+         }
+
+         @Override
+         public boolean verifyEvidence(final Evidence evidence) {
+            for (Credential credential : accountEntry.getCredentials()) {
+               if (credential != null || credential.canVerify(evidence)) {
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java:142]: Will produce  a NPE. I assume it should be `&&`.

File: server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+               realmName = defaultRealm;
+            } else {
+               throw log.noRealmFoundInProperties();
+            }
+         }
+      }
+
+      // users, which are in groups file only
+      for (String userName : groups.stringPropertyNames()) {
+         if (accounts.containsKey(userName) == false) {
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java:297]: `!accounts.containsKey(userName)`

File: server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+
+         return realm;
+      }
+
+   }
+
+   private static class LoadedState {
+
+      private final Map<String, AccountEntry> accounts;
+      private final String realmName;
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java:433]: `realmName` is never used.

File: server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java [hunk 9]
```diff
@@ -0,0 +1,500 @@
...
+         String[] groupArray = groups.split(",");
+         Set<String> groupsSet = new HashSet<>(groupArray.length);
+         for (String current : groupArray) {
+            String value = current.trim();
+            if (value.length() > 0) {
+               groupsSet.add(value);
+            }
+         }
+
+         return Collections.unmodifiableSet(groupsSet);
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java:482]: Can be simplified to:

```java
         return Arrays.stream(groups.split(","))
               .map(String::trim)
               .filter(s -> !s.isEmpty())
               .collect(Collectors.collectingAndThen(Collectors.toSet(), Collections::unmodifiableSet));
```

File: server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java [hunk 10]
```diff
@@ -0,0 +1,500 @@
...
+         return credentials;
+      }
+
+
+      public Set<String> getGroups() {
+         return groups;
+      }
+   }
+
+
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java:499]: Unnecessary whitespace.

File: server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java (hunk 8 above)
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/security/realm/EncryptedPropertiesSecurityRealm.java:420]: I'd rather keep it around in case we may want to use it. It's code that comes from Elytron's original implementation.

**Comment by tristantarrant** [comment]
@ryanemerson I've made the requested changes, aside from one which I commented on above.

**Comment by ryanemerson** [comment]
@tristantarrant Are these failures related:

```
org.infinispan.server.persistence.ManagedConnectionOperations.testPutGetRemove
org.infinispan.server.persistence.ManagedConnectionOperations.testTwoCachesSameCacheStore
```

They appear on all three runs, but not on our latest master build.

**Comment by ryanemerson** [comment]
@tristantarrant Are these failures related:

```
org.infinispan.server.persistence.ManagedConnectionOperations.testPutGetRemove
org.infinispan.server.persistence.ManagedConnectionOperations.testTwoCachesSameCacheStore
```

They appear on all three runs, but no on our latest master builds

**Comment by ryanemerson** [comment]
I saw the failures on other PRs which answered the question. Thanks @tristantarrant
